pywr-core/src/aggregated_node.rs

```diff
@@ -263,7 +337,71 @@ fn get_norm_proportional_factor_pairs(
         .collect::<Vec<_>>()
 }
 
-/// Ratio factors
+/// Calculate constant factor pairs for proportional factors.
+///
+/// There should be one less factor than node indices. The factors correspond to each of the node
+/// indices after the first. Factor pairs relating the first index to each of the other indices are
+/// calculated. This requires the sum of the factors to be greater than 0.0 and less than 1.0. If
+/// any of the factors are not constant, the factor pairs will contain `None` values.
+fn get_const_norm_proportional_factor_pairs(
+    factors: &[MetricF64],
+    nodes: &[NodeIndex],
+    values: &ConstParameterValues,
+) -> Vec<NodeConstFactorPair> {
+    if factors.len() != nodes.len() - 1 {
+        panic!("Found {} proportional factors and {} nodes in aggregated node. The number of proportional factors should equal one less than the number of nodes.", factors.len(), nodes.len());
+    }
+
+    // First get the current factor values
+    let values: Vec<Option<f64>> = factors
+        .iter()
+        .map(|f| f.try_get_constant_value(values))
+        .collect::<Result<Vec<_>, PywrError>>()
+        .expect("Failed to get current factor values.");
+
+    let n0 = nodes[0];
+
+    // To calculate the factors we require that every factor is available.
+    if values.iter().any(|v| v.is_none()) {
+        // At least one factor is not available; therefore we can not calculate "f0"
+        nodes
+            .iter()
+            .skip(1)
+            .zip(values)
+            .map(move |(&n1, f1)| {
+                NodeConstFactorPair::new(NodeConstFactor::new(n0, None), NodeConstFactor::new(n1, f1))
+            })
+            .collect::<Vec<_>>()
+    } else {
+        // All factors are available; therefore we can calculate "f0"
+        // TODO do we need to assert that each individual factor is positive?
```

Comment:
to do in this PR? Should it be a general requirement that factors are non-negative checked in the schema?

Reply:
This has been done.